src/main/java/org/apache/commons/text/similarity/SorensenDicesSimilarity.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.text.similarity;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ *
+ * @since 1.7
+ */
+public class SorensenDicesSimilarity implements SimilarityScore<Double> {
+
+    /**
+     * @param left  the first CharSequence, must not be null
+     * @param right the second CharSequence, must not be null
+     * @return result similarity
+     * @throws IllegalArgumentException if either CharSequence input is {@code null}
+     */
+
+    @Override
+    public Double apply(final CharSequence left, final CharSequence right) {
+
+        if (left == null || right == null) {
+            throw new IllegalArgumentException("CharSequences must not be null");
+        }
+
+        if (left.equals(right)) {
+            return 1d;
+        }
+
+        if ("".equals(left) || "".equals(right)) {
```

Comment:
Perhaps:
```
if (Math.min(left.length(), right.length()) < 2) {
    // No bigrams
    return 0;
}
```

Reply:
I think existing logic is better, in your logic length of both strings have to calculate, then Math.min
in the existing best case is if ```left``` is empty then return.

Reply:
I'm not familiar with a `CharSequence` which has an expensive calculation for its length. Note that this logic will avoid all the `Set` creation if one of the inputs is a single character.

Reply:
Also note that you are calling `String.equals(Object)` which only compares to another `String`. In this case if the argument **is** a `String` then `String.length()` will be faster logic. For any other `CharSequence` this logic will fail as it will be ignored by `String.equals(Object)`.

You should look at the JDK 1.8 source code for `String.contentEquals(CharSequence)`. Then decide how fast that will be compared to known implementations of `length()` in JDK 1.8 [CharSequence](https://docs.oracle.com/javase/8/docs/api/java/lang/CharSequence.html) or the commons-text `TextStringBuilder.length()`. Most implementations I found just return an `int`. `CharBuffer` returns the computation of one `int` minus another `int`.

Reply:
well, I wanted to validate that claim so I wrote small jmh script and here are the result.

- https://github.com/ameyjadiye/jmh-sampling
- [SamplingEqVsMath.java](https://github.com/ameyjadiye/jmh-sampling/blob/master/src/main/java/com/codeinventory/jmh_sampling/SamplingEqVsMath.java)

quick readability of the program goes down over the point of latency, I'm ok to make that change.

Reply:
Not a good benchmark since you only tested empty strings and the same string at that. This is testing the reference is the same (effectively String.equals(String s) { return this == s; }) and not even looking at the characters in the string.

What about strings that are 10, 20, 100+ characters long?

Plus what about CharSequence where String.equals will not work? Did you look at String.contentEquals?

Reply:
Plus if I understand JMH correctly then it will optimise out all the code in your benchmark. You need to return something from the methods or use a BlackHole to accept the boolean result of the test.

Reply:
> quick readability of the program goes down over the point of latency, I'm ok to make that change but with some change as **< 2** looks bit wrong condition ...

Then you make it more readable:

```
/** The minimum length for a CharSequence to contain a bigram. */
private final int MIN_BIGRAM_LENGTH = 2;


if (left.length() < MIN_BIGRAM_LENGTH || right.length() < MIN_BIGRAM_LENGTH) {
    // No bigrams
    return 0d;
}
```

Reply:
what if `left="a"` and `right="a"` , test will fail as a=a means it expects 100% similarity. returning 0 is not good. we are just checking whether args passed are empty or not, let algorithm decide further what do with charSequence. WDYT ?

Reply:
The line above the length comparison has an attempt at testing for equality:

```
if (left.equals(right)) {
    return 1d;
}
```

So the code will not reach the length check anyway. 

However I note that the use of `equals` is actually invalid. It is not defined in the `CharSequence` interface and so comparing a `String` with a `StringBuilder` would fail even if they have the same chars because `String.equals` only matches other `String` objects.

Note also that this is an algorithm that does comparisons using at least two characters. So inputting a single character is not valid for the main algorithm processing. It needs to be handled as an edge case.

So handling char sequences of a single character should return 1 or 0 (because there are no bigrams). It makes more sense to return 1. The decision needs to be in the Javadoc. I would vote for returning 1.

Here are the edge cases and proposed results:

```
null   vs null         IllegalArgumentException
""     vs ""           1
"a"    vs "a"          1
"ab"   vs "ab"         1
"abc"  vs "abd"        0.5   { [ab][bc]   verses   [ab][bd] = 2 * 1 / (2 + 2) }
```



So you need something like:

```
if (left == null || right == null) {
    throw new IllegalArgumentException("CharSequences must not be null");
}

// Edge case for equality. Use Commons-Lang StringUtils which is a project dependency.
if (StringUtils.equals(left, right)) {
    return 1d;
}

// Edge case for input not analysed by the algorithm
if (left.length() <= 2 || right.length() <= 2) {
    // No bigrams
    return 0d;
}
```

Reply:
I think the last comment will generate a by-product of this pull request/issue by @aherbert , which is the improvement for the `StringUtils` equality check functions :-) 

I am not too concerned if our performance is excellent here at this moment. As long as it is not extremely slow, I think it should be fine to settle for a solution that is easy to read & maintain, and performs OK. 

Once it's working, we can fine-tune, especially after some user feedback :+1: but good discussion.

Reply:
@aherbert , thanks, got your point. after adding some more test cases I came to know the missing handles, changed the code.